File: COGS/983.cpp
#include <iostream>
#include <cstring>
#include <cstdio>
#define MAXN 55
#define MAX true
#define MIN false
#define INF 1<<16
#define SpeedUp ios::sync_with_stdio(false)
#define FILE
//#define Debug
using namespace std;

int N,M;
int Max,Min;
int f[MAXN];
int Last[MAXN][MAXN],Now[MAXN][MAXN];

inline int total(int i,int j){
	return (i<=j) ? (f[j]-f[i-1]) : (f[N]-f[i-1]+f[j]);
}

void init()
{
	int r;
	SpeedUp;
	#ifdef FILE
	freopen("numgame.in","r",stdin);
	freopen("numgame.out","w",stdout);
	#endif
	cin>>N>>M;
	memset(f,0,sizeof(f));
	for(int i=1;i<=N;i++)
	{
		cin>>r;
		f[i]=f[i-1]+r;
	}
}

int Dp(bool LHD,int &Ans)
{
	Ans=((LHD) ? 0 : INF);			
	int dis,t1,t2,extre;
	dis=(LHD ? -1 : INF);
	#ifdef Debug
	cout<<"dis = "<<dis<<endl
		<<"Ans = "<<Ans<<endl;
	#endif
	for(int i=1;i<=N;i++)
		for(int j=1;j<=N;j++)
		{
			Last[i][j]=((i==j) ? 1 : dis);
		}
	
	#ifdef Debug
	cout<<"Last init:"<<endl;
	for(int i=1;i<=N;i++)
	{
		for(int j=1;j<=N;j++)
			cout<<Last[i][j]<<"\t";
		cout<<endl;
	}
	#endif
	for(int L=1;L<=M-1;L++)
	{
		for(int i=1;i<=N;i++)
			for(int j=1;j<=N;j++)
				Now[i][j]=dis;
		for(int i=1;i<=N;i++)
			for(int j=i+L-1;j<=N+i-1;j++)
			{
				t1=(j-1)%N+1;
				for(int k=i+L-1;k<=j;k++)
				{
					t2=(k-1)%N+1;
					if(Last[i][t2]==-1 || Last[i][t2]==INF)
						continue;
					extre=total(t2,t1);
					extre=Last[i][t2]*((10+extre%10)%10);
					#ifdef Debug
					cout<<"-----Extre is "<<extre<<endl;
					#endif
					if(!LHD)
						if(extre<Now[i][t1])
							Now[i][t1]=extre;
					if(LHD)
						if(extre>Now[i][t1])
							Now[i][t1]=extre;
				}
			}
		#ifdef Debug
		cout<<"Last:"<<endl;
		for(int i=1;i<=N;i++){
			for(int j=1;j<=N;j++)
				cout<<Last[i][j]<<"\t";
			cout<<endl;
		}
		cout<<"Now:"<<endl;
		for(int i=1;i<=N;i++){
			for(int j=1;j<=N;j++)
				cout<<Now[i][j]<<"\t";
			cout<<endl;
		}
		#endif
		memcpy(Last,Now,N);
		for(int i=1;i<=N;i++)
			for(int j=1;j<=N;j++)
				Last[i][j]=Now[i][j];
		#ifdef Debug
		cout<<"Last = Now = :"<<endl;
		for(int i=1;i<=N;i++){
			for(int j=1;j<=N;j++)
				cout<<Last[i][j]<<"\t";
			cout<<endl;
		}
		#endif
	}
	
	for(int i=1;i<=N;i++)
		for(int j=1;j<=N;j++)
			if(i!=j && Now[i][j]!=INF && Now[i][j]!=-1)
			{
				extre=total(j%N+1,(i+N-2)%N+1);
				extre=Now[i][j]*((10+extre%10)%10);
				if(LHD)
					if(extre>Ans)
						Ans=extre;
				if(!LHD)
					if(extre<Ans)
						Ans=extre;
			}
	return Ans;
}

void work()
{
	int Ans;
	#ifdef Debug
	cout<<"Min Dp:"<<endl;
	#endif
	cout<<Dp(MIN,Ans)<<endl;
	
	//Min=Ans;
	#ifdef Debug
	cout<<"--------------------"<<endl
		<<"Max Dp:"<<endl;
	#endif
	cout<<Dp(MAX,Ans)<<endl;
	//Max=Ans;
}

int main()
{
	init();
	work();
	return 0;
}

File: COGS/983.pas
const   infns                   = 'numgame.in';
        outfns                  = 'numgame.out';
        maxn                    =  50;
var     m, n                    :  byte;
        max, min                :  longint;
        a                       :  array[0 .. maxn] of longint;
        last, now               :  array[1 .. maxn, 1 .. maxn] of longint;

procedure out;
  begin
    assign(output, outfns);
    rewrite(output);
    writeln(min);
    writeln(max);
    close(output)
  end;  { end of out }

function  sum(i, j : byte) : longint;
  begin
    if i <= j then sum := a[j] - a[i-1]
              else sum := a[n] - a[i-1] + a[j];
  end;  { end of sum }

procedure catch(kind : byte;  var  ans : longint);
var  i, j1, j, k1, k, l     : integer;
     best, dis              : longint;
  begin
    case kind of
      0 : dis := maxlongint;
      1 : dis := -1;
    end;
    for i := 1 to n do
      for j := 1 to n do last[i,j] := dis;
    for i := 1 to n do last[i,i] := 1;
    
    for l := 1 to m - 1 do
      begin
        for i := 1 to n do
          for j := 1 to n do 
		  	now[i,j] := dis;
        for i := 1 to n do
          for j1 := i + l - 1 to n + i - 1 do
            begin
              j := (j1 - 1) mod n + 1;
              for k1 := i + l - 1 to j1 do
                begin
                  k := (k1 - 1) mod n + 1;
                  if (last[i,k] = -1) or (last[i,k] = maxlongint) then continue;
                  best := sum(k, j);
                  best := (10 + best mod 10) mod 10;
                  best := last[i,k] * best;
                  case kind of
                    0 : if best < now[i,j] then now[i,j] := best;
                    1 : if best > now[i,j] then now[i,j] := best;
                  end;
                end;{k1}
            end;{i}
        last := now;
      end;
    case kind of
      0 :  ans := maxlongint;
      1 :  ans := 0;
    end;
    for i := 1 to n do
      for j := 1 to n do
        if (i <> j) and (now[i,j] <> maxlongint) and (now[i,j] <> -1) then
          begin
            best := sum(j mod n + 1, (i+n-2) mod n+1);
            best := (10 + best mod 10) mod 10;
            best := now[i,j] * best;
            case kind of
              0 : if best < ans then ans := best;
              1 : if best > ans then ans := best;
            end;
          end
  end;  { end of catch }

procedure main;
var  ans                : longint;
  begin
    catch(0, ans);
    min := ans;
    catch(1, ans);
    max := ans;
  end;  { end of main }

procedure init;
var  i, x               : integer;
  begin
    assign(input, infns);
    reset(input);
    readln(n, m);
    fillchar(a, sizeof(a), 0);
    for i := 1 to n do
      begin
        readln(x);
        a[i] := a[i-1] + x;
      end;
    close(input);
  end;  { end of init }

begin
  init;
  main;
  out;
end.
